src/deltalake_ext.rs

```diff
@@ -0,0 +1,322 @@
+    arrow_writer: ArrowWriter<InMemoryWriteableCursor>,
+    partition_values: HashMap<String, String>,
+    num_records: i64,
+}
+
+/// A writer that writes record batches in parquet format to a table location.
+/// This should be used along side a DeltaTransaction wrapping the same DeltaTable instance.
+impl DeltaParquetWriter {
+    /// Initialize the writer from the given table path and delta schema
+    pub async fn for_table_path_with_schema(table_path: String, schema: &Schema) -> Result<DeltaParquetWriter, DeltaWriterError> {
+        // Initialize storage backend for the given table path
+        let storage = deltalake::get_backend_for_uri(table_path.as_str())?;
+
+        // Initialize an arrow schema ref from the delta table schema
+        let arrow_schema = <ArrowSchema as From<&Schema>>::from(schema);
+        let arrow_schema_ref = Arc::new(arrow_schema);
+
+        // Initialize writer properties for the underlying arrow writer
+        let writer_properties = WriterProperties::builder()
+            // TODO: Extract config/env for writer properties and set more than just compression
+            .set_compression(Compression::SNAPPY)
+            .build();
+
+        let cursor = InMemoryWriteableCursor::default();
+        // TODO: Handle error
+        let arrow_writer = ArrowWriter::try_new(cursor.clone(), arrow_schema_ref.clone(), Some(writer_properties.clone())).unwrap();
+
+        Ok(Self {
+            table_path,
+            storage,
+            arrow_schema_ref,
+            writer_properties,
+            cursor,
+            arrow_writer,
+            partition_values: HashMap::new(),
+            num_records: 0,
+        })
+    }
+
+    /// Writes the record batch in-memory and updates internal state accordingly.
+    /// This method buffers the write stream internally so it can be invoked for many record batches and flushed after the appropriate number of bytes has been written.
+    pub async fn write_record_batch(&mut self, partition_cols: &Vec<String>, record_batch: &RecordBatch) -> Result<(), DeltaWriterError> {
+        let partition_values = extract_partition_values(partition_cols, record_batch)?;
+
+        // TODO: Verify that this RecordBatch contains partition values that agree with previous partition values (if any) for the current writer context
+        self.partition_values = partition_values;
+
+        // TODO: collect/update column stats
+
+        // write the record batch to the held arrow writer
+        self.arrow_writer.write(record_batch)?;
+
+        // increment num records (for the entire file) by the number of records in the handled batch
+        self.num_records += record_batch.num_rows() as i64;
+
+        Ok(())
+    }
+
+    /// Returns the current byte length of the in memory buffer.
+    /// This should be used by the caller to decide when to call `complete_file`.
+    pub fn buffer_len(&self) -> Result<usize, DeltaWriterError> {
+        Ok(self.cursor.data().len())
```

Comment:
do we really need to wrap this with result?

Reply:
Nope. Fixed in https://github.com/delta-io/kafka-delta-ingest/pull/1/commits/c90ff3418c7bbf0e56cce7a7dda990e3c616140a#diff-5d8c9545386ed0b9dcf9dab1722a6b484e1d3014d9a2470788dd7b20509ca642R159.